internal/kldtx/addressbook.go

```diff
@@ -0,0 +1,184 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kldtx
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/kaleido-io/ethconnect/internal/kldeth"
+	"github.com/kaleido-io/ethconnect/internal/kldutils"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	defaultRPCEndpointProp = "endpoint"
+)
+
+// AddressBook looks up RPC URLs based on a remote registry, and optionally
+// resolves hostnames to IP addresses using a hosts file
+type AddressBook interface {
+	lookup(addr string) (kldeth.RPCClient, error)
+}
+
+// AddressBookConf configuration
+type AddressBookConf struct {
+	kldutils.HTTPRequesterConf
+	AddressbookURLPrefix string                   `json:"urlPrefix"`
+	HostsFile            string                   `json:"hostsFile"`
+	PropNames            AddressBookPropNamesConf `json:"propNames"`
+}
+
+// AddressBookPropNamesConf configures the JSON property names to extract from the GET response on the API
+type AddressBookPropNamesConf struct {
+	RPCEndpoint string `json:"endpoint"`
+}
+
+// NewAddressBook construtor
+func NewAddressBook(conf *AddressBookConf, rpcConf *kldeth.RPCConf) AddressBook {
+	ab := &addressBook{
+		conf:                conf,
+		fallbackRPCEndpoint: rpcConf.RPC.URL,
+		hr:                  kldutils.NewHTTPRequester("Addressbook", &conf.HTTPRequesterConf),
+		addrToHost:          make(map[string]string),
+		hostToRPC:           make(map[string]kldeth.RPCClientAll),
+	}
+	propNames := &conf.PropNames
+	if propNames.RPCEndpoint == "" {
+		propNames.RPCEndpoint = defaultRPCEndpointProp
+	}
+	if ab.conf.AddressbookURLPrefix != "" && !strings.HasSuffix(ab.conf.AddressbookURLPrefix, "/") {
+		ab.conf.AddressbookURLPrefix += "/"
+	}
+	return ab
+}
+
+type addressBook struct {
+	conf                *AddressBookConf
+	hr                  *kldutils.HTTPRequester
+	mtx                 sync.Mutex
+	fallbackRPCEndpoint string
+	addrToHost          map[string]string
+	hostToRPC           map[string]kldeth.RPCClientAll
+}
+
+// testRPC uses a simple net_version JSON/RPC call to test the health of a cached connection
+func (ab *addressBook) testRPC(rpc kldeth.RPCClient) bool {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	var netID string
+	if err := rpc.CallContext(ctx, &netID, "net_version"); err != nil {
+		log.Errorf("Stale RPC connection in pool: %s", err)
+		return false
+	}
+	return true
+}
+
+func (ab *addressBook) resolveHost(endpoint string) (*url.URL, error) {
+	// We need to connect, but first we need to resolve the hostname to an IP
+	url, err := url.Parse(endpoint)
+	if err != nil {
+		log.Errorf("Invalid URL '%s': %s", endpoint, err)
+		return nil, fmt.Errorf("Invalid URL obtained for address")
+	}
+
+	if ab.conf.HostsFile != "" {
```

Comment:
For single region environments, we are not going to find the hosts file and so we use the endpoint URL as is, correct? (without translating to an IP address)